tempodb/encoding/vparquet/wal_block.go

```diff
@@ -0,0 +1,577 @@
+		if err != nil {
+			return nil, nil, fmt.Errorf("error opening file: %w", err)
+		}
+
+		b.flushed = append(b.flushed, &walBlockFlush{
+			file: pf,
+			ids:  common.NewIDMap(),
+		})
+		b.flushedSize += sz
+	}
+
+	// iterate through all files and build meta
+	for i, page := range b.flushed {
+		iter := makeIterFunc(context.Background(), page.file.RowGroups(), page.file)(columnPathTraceID, nil, columnPathTraceID)
+		defer iter.Close()
+
+		for {
+			match, err := iter.Next()
+			if err != nil {
+				return nil, nil, fmt.Errorf("error opening wal folder [%s %d]: %w", b.meta.BlockID.String(), i, err)
+			}
+			if match == nil {
+				break
+			}
+
+			for _, e := range match.Entries {
+				switch e.Key {
+				case columnPathTraceID:
+					traceID := e.Value.ByteArray()
+					b.meta.ObjectAdded(traceID, 0, 0)
+					page.ids.Set(traceID)
+				}
+			}
+		}
+	}
+
+	return b, nil, nil
+}
+
+// createWALBlock creates a new appendable block
+func createWALBlock(id uuid.UUID, tenantID string, filepath string, _ backend.Encoding, dataEncoding string, ingestionSlack time.Duration) (common.WALBlock, error) {
+	b := &walBlock{
+		meta: &backend.BlockMeta{
+			Version:  VersionString,
+			BlockID:  id,
+			TenantID: tenantID,
+		},
+		path:           filepath,
+		ids:            common.NewIDMap(),
+		ingestionSlack: ingestionSlack,
+	}
+
+	// build folder
+	err := os.MkdirAll(b.walPath(), os.ModePerm)
+	if err != nil {
+		return nil, err
+	}
+
+	dec, err := model.NewObjectDecoder(dataEncoding)
+	if err != nil {
+		return nil, err
+	}
+	b.decoder = dec
+
+	return b, nil
+}
+
+func ownsWALBlock(entry fs.DirEntry) bool {
+	// all vParquet wal blocks are folders
+	if !entry.IsDir() {
+		return false
+	}
+
+	_, _, version, err := parseName(entry.Name())
+	if err != nil {
+		return false
+	}
+
+	return version == VersionString
+}
+
+type walBlockFlush struct {
+	file *parquet.File
+	ids  *common.IDMap
+}
+
+type walBlock struct {
+	meta           *backend.BlockMeta
+	path           string
+	ingestionSlack time.Duration
+
+	// Unflushed data
+	traces        []*Trace
+	ids           *common.IDMap
+	unflushedSize int64
+
+	// Flushed data
+	flushed     []*walBlockFlush
+	flushedSize int64
+
+	writer  *parquet.GenericWriter[*Trace]
+	decoder model.ObjectDecoder
+}
+
+func (b *walBlock) BlockMeta() *backend.BlockMeta {
+	return b.meta // jpe make ingestion slack a handled by BlockMeta
+}
+
+func (b *walBlock) Append(id common.ID, buff []byte, start, end uint32) error {
+	// if decoder = nil we were created with OpenWALBlock and will not accept writes
+	if b.decoder == nil {
+		return nil
+	}
+
+	trace, err := b.decoder.PrepareForRead(buff)
+	if err != nil {
+		return fmt.Errorf("error preparing trace for read: %w", err)
+	}
+
+	tr := traceToParquet(id, trace, getPooledTrace())
+
+	start, end = b.adjustTimeRangeForSlack(start, end, 0)
+
+	b.meta.ObjectAdded(id, start, end)
+
+	// add to current
+	b.traces = append(b.traces, tr)
+	b.ids.Set(id)
+
```

Comment:
i think we need to return `tr` to the pool here. profiling this branch in ops is showing 57% of all allocated memory in the `traceToParquet` conversion code.